### ISSUE TITLE
proposal: Go 2: adapt mips_siflower

### ISSUE THREAD

**Issue by wangjony** [golang/go#61254]
GL.iNet GL-SFT1200 WiFi Router
Architecture: mips_siflower
Operating System: OpenWrt
The GL.iNet GL-SFT1200 comes with an Siflower SF19A28 mips_siflower processor,128 MB flash, and 128 MB RAM. Hope to adapt mips_siflower，Thank you so much

**Comment by seankhliao** [comment]
What exactly is being proposed here?

**Comment by wangjony** [comment]
> What exactly is being proposed here?
Support for mips_siflower architecture🙏🏻🙏🏻🙏🏻

**Comment by rittneje** [comment]
@bcmills I think you meant "GOARCH settings" not "GOOS" settings. (Just to avoid confusion.)

**Comment by bcmills** [comment]
Indeed; edited my previous comment.

**Comment by bcmills** [comment]
That suggest that support for this chip would be considered a new port for Go. See https://go.dev/wiki/PortingPolicy#requirements-for-a-new-port for what is required to add one.

**Comment by Rongronggg9** [comment]
### [`/proc/cpuinfo`](https://www.cnblogs.com/osnosn/p/16686059.html)
```
system type             : MIPS sf19a28
machine                 : GL-SFT1200
processor               : 0
cpu model               : MIPS interAptiv (multi) V2.8  FPU V0.0
BogoMIPS                : 531.66
wait instruction        : yes
microsecond timers      : yes
tlb_entries             : 64
extra interrupt vector  : yes
hardware watchpoint     : yes, count: 4, address/irw mask: [0x0ffc, 0x0ffc, 0x0ffb, 0x0ffb]
isa                     : mips1 mips2 mips32r1 mips32r2
ASEs implemented        : mips16 dsp dsp2 mt eva
```

Datasheet: https://www.siflower.com.cn/assets/upload/file/SF19A2890_datasheet.pdf

Its CPU is MIPS32® interAptiv, a MIPS32r2 CPU.

### readelf

#### [busybox_1.29.3-3_mips_siflower.ipk](https://fw.gl-inet.com/releases/v18.06.5/packages-3.6/siflower/packages/busybox_1.29.3-3_mips_siflower.ipk)

```
ELF Header:
  Magic:   7f 45 4c 46 01 01 01 00 01 00 00 00 00 00 00 00
  Class:                             ELF32
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       1
  Type:                              EXEC (Executable file)
  Machine:                           MIPS R3000
  Version:                           0x1
  Entry point address:               0x403760
  Start of program headers:          52 (bytes into file)
  Start of section headers:          0 (bytes into file)
  Flags:                             0x1005, noreorder, cpic, o32, mips1
  Size of this header:               52 (bytes)
  Size of program headers:           32 (bytes)
  Number of program headers:         9
  Size of section headers:           0 (bytes)
  Number of section headers:         0
  Section header string table index: 0
...
      [Requesting program interpreter: /lib/ld-musl-mipsel.so.1]
...
 0x00000001 (NEEDED)                     Shared library: [libgcc_s.so.1]
 0x00000001 (NEEDED)                     Shared library: [libc.so]
```

MIPS I (weird!) little-endian. `ld.so` is `/lib/ld-musl-mipsel.so.1`.

#### [busybox_1.28.4-3_mipsel_24kc.ipk](https://archive.openwrt.org/releases/packages-18.06/mipsel_24kc/base/busybox_1.28.4-3_mipsel_24kc.ipk)
```
ELF Header:
  Magic:   7f 45 4c 46 01 01 01 00 01 00 00 00 00 00 00 00
  Class:                             ELF32
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       1
  Type:                              EXEC (Executable file)
  Machine:                           MIPS R3000
  Version:                           0x1
  Entry point address:               0x403940
  Start of program headers:          52 (bytes into file)
  Start of section headers:          0 (bytes into file)
  Flags:                             0x74001005, noreorder, cpic, o32, mips16, mips32r2
  Size of this header:               52 (bytes)
  Size of program headers:           32 (bytes)
  Number of program headers:         9
  Size of section headers:           0 (bytes)
  Number of section headers:         0
  Section header string table index: 0
...
      [Requesting program interpreter: /lib/ld-musl-mipsel-sf.so.1]
...
 0x00000001 (NEEDED)                     Shared library: [libgcc_s.so.1]
 0x00000001 (NEEDED)                     Shared library: [libc.so]
```

MIPS32r2 little-endian. `ld.so` is `/lib/ld-musl-mipsel-sf.so.1`.

<hr>

The reason why one cannot "execute an ELF built for `mipsel_24kc`" is clear now:
1. OpenWrt opkg refuses to install `mipsel_24kc` ipk as is not considered a compatible architecture.
2. A dynamically linked ELF is unable to find a proper `ld.so`.

In theory, a static linked Go program should be able to execute on `mips_siflower`. So @wangjony, could you test?

In conclusion, I don't think it need a new port for Go to support `mips_siflower`, but the downstream OpenWrt GL.iNet provides is a bit "toxic". No much things can do from our side, but users needs to configure CGO properly.

**Comment by rsc** [comment]
Is the only difference the dynamic linker that's embedded in the binary?

**Comment by rsc** [comment]
This proposal has been added to the [active column](https://go.dev/s/proposal-status#active) of the proposals project
and will now be reviewed at the weekly proposal review meetings.
— rsc for the proposal review group

**Comment by iGerman00** [comment]
> In theory, a static linked Go program should be able to execute on `mips_siflower`

I can confirm that works.
[Example of building a working binary for that platform](https://gist.github.com/iGerman00/b45f78685f7db0fa264edf7a0d81bd50)

**Comment by ianlancetaylor** [comment]
@wangjony Is there anything to do here?  Thanks.

**Comment by rsc** [comment]
It sounds like if you want to compile for siflower from non-siflower systems you can do that by using static binaries but also by using dynamic binaries with `go build -ldflags=-I=/lib/ld-musl-mipsel-sf.so.1`. Do I have that correct?

**Comment by Rongronggg9** [comment]
> It sounds like if you want to compile for siflower from non-siflower systems you can do that by using static binaries
 
Yes.

> but also by using dynamic binaries with `go build -ldflags=-I=/lib/ld-musl-mipsel-sf.so.1`.

Nope. It is `/lib/ld-musl-mipsel.so.1`. `sf` means soft-float instead of siflower. A `mips-siflower` downstream OpenWrt installation comes without `ld-musl-mipsel-sf.so.1` as the SoC is of an FPU.

I think we've proved that the proposal is a mistake.

**Comment by rsc** [comment]
Based on the discussion above, this proposal seems like a **[likely decline](https://go.dev/s/proposal-status#likely-decline)**.
— rsc for the proposal review group

**Comment by rsc** [comment]
No change in consensus, so **[declined](https://go.dev/s/proposal-status#declined)**.
— rsc for the proposal review group
